sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/DigitalTwinMetadataJsonConverter.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Azure.DigitalTwins.Core
+{
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+
+    /// <summary>
+    /// JSON converter to make it easier to deserialize a <see cref="BasicDigitalTwin"/>.
+    /// </summary>
+    public class DigitalTwinMetadataJsonConverter : JsonConverter<DigitalTwinMetadata>
+    {
+        /// <inheritdoc/>
+        public override DigitalTwinMetadata Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
+
+            if (reader.TokenType == JsonTokenType.StartObject)
+            {
+                var metadata = new DigitalTwinMetadata();
+                reader.Read(); // Advance into our object.
+
+                while (reader.TokenType != JsonTokenType.EndObject) // Until we reach the end of the object we began reading
+                {
+                    if (reader.TokenType == JsonTokenType.PropertyName)
+                    {
+                        if (reader.GetString() == DigitalTwinsJsonPropertyNames.MetadataModel)
+                        {
+                            reader.Read(); // advance to the string
+                            metadata.ModelId = JsonSerializer.Deserialize<string>(ref reader, options);
+                            reader.Read(); // Consume the string we deserialized
+                        }
+                        else
+                        {
+                            var propertyName = reader.GetString();
+                            reader.Read(); // Should place us on StartObject
+                            if (reader.TokenType == JsonTokenType.StartObject)
+                            {
+                                metadata.PropertyMetadata[propertyName] = JsonSerializer.Deserialize<DigitalTwinPropertyMetadata>(ref reader, options);
+                                reader.Read(); // Consume EndObject
+                            }
+                        }
+                    }
+                }
+
+                return metadata;
+            }
+
+            throw new Exception($"Unknown token type {reader.TokenType} at index {reader.TokenStartIndex}.");
```

Comment:
Recommend a better exception type. I don't know of one off hand. I wish JsonConvert base class had some kind of "ReportError" functionality which would generate an exception (with information like the current token, offset, etc) on behalf of custom converters.

Reply:
I looked at a bunch and don't see an obvious one. @pakrym, any idea here?

Reply:
Don't think there is an error reporting method but you can always throw JsonException(https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonexception?view=netcore-3.1).

Reply:
https://github.com/dotnet/runtime/blob/c08be254963194f023c2e76867093a300ff713fe/src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs

This is what is used in the built-in converters.

Reply:
What do you think, `InvalidOperationException`?

Reply:
I'm in favor of throwing JsonException, just like the built-in converters. Unfortunately the documentation for writing a custom JsonConverter doesn't describe a best practice here... If we throw a JsonException at least a caller would just be able to handle that single type for all JsonExceptions (including a completely malformed payload?)